File: security-assessment-dashboard/backend/schemas/finding.py
"""Request/response schemas for the Correlation Engine's Findings API.

Every resource here is read-only from the API's perspective, except
triggering a correlation run itself — findings are only ever written by
:class:`backend.services.correlation_service.CorrelationService`.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel

from backend.models.enums import FindingConfidence, FindingReferenceType, FindingSeverity, FindingStatus
from backend.schemas.host_inventory import ExecutionHistoryEntryRead, HostSummaryRead, ObservationRead, ServiceRead


class FindingReferenceRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    reference_type: FindingReferenceType
    reference_value: str


class FindingEvidenceRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    source_tool: str
    title: str | None
    content: str | None
    file_path: str | None
    created_at: datetime


class FindingSummaryRead(BaseModel):
    """Lightweight shape for the Findings list page."""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    assessment_id: uuid.UUID
    host_id: uuid.UUID | None
    rule_id: str
    plugin: str | None
    title: str
    severity: FindingSeverity
    confidence: FindingConfidence
    category: str | None
    status: FindingStatus
    first_seen: datetime
    last_seen: datetime
    host_label: str | None = None
    evidence_count: int = 0
    observation_count: int = 0


class FindingDetailRead(BaseModel):
    """Full Finding Details page shape."""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    assessment_id: uuid.UUID
    host_id: uuid.UUID | None
    source_execution_id: uuid.UUID | None
    rule_id: str
    plugin: str | None
    title: str
    description: str | None
    impact: str | None
    severity: FindingSeverity
    confidence: FindingConfidence
    category: str | None
    cvss_score: float | None
    cwe: str | None
    owasp: str | None
    remediation: str | None
    status: FindingStatus
    first_seen: datetime
    last_seen: datetime
    created_at: datetime
    updated_at: datetime
    host: HostSummaryRead | None = None
    affected_services: list[ServiceRead] = []
    supporting_observations: list[ObservationRead] = []
    evidence: list[FindingEvidenceRead] = []
    references: list[FindingReferenceRead] = []
    execution_history: list[ExecutionHistoryEntryRead] = []


class CorrelationRunRead(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    assessment_id: uuid.UUID | None
    status: str
    started_at: datetime
    completed_at: datetime | None
    hosts_evaluated: int
    rules_evaluated: int
    findings_created: int
    findings_updated: int
    error_message: str | None


class CorrelationRunRequest(BaseModel):
    assessment_id: uuid.UUID | None = None


class CorrelationRunResultRead(BaseModel):
    """Result of a just-completed correlation run -- real counts, not an estimate."""

    assessment_id: uuid.UUID | None
    hosts_evaluated: int
    rules_evaluated: int
    rule_count: int
    findings_created: int
    findings_updated: int


class CorrelationStatusRead(BaseModel):
    """The most recent correlation run's outcome, plus how many rules are registered."""

    registered_rule_count: int
    last_run: CorrelationRunRead | None
    recent_runs: list[CorrelationRunRead] = []

File: security-assessment-dashboard/frontend/src/types/finding.ts
import type { ExecutionHistoryEntry, HostSummary, Observation, Service } from "@/types/host-inventory";

export type FindingSeverity = "critical" | "high" | "medium" | "low" | "info";
export type FindingConfidence = "confirmed" | "high" | "medium" | "low";
export type FindingStatus = "open" | "confirmed" | "false_positive" | "accepted_risk" | "remediated" | "duplicate";
export type FindingReferenceType = "cwe" | "owasp" | "capec" | "cve" | "vendor_url" | "documentation_url";

export const FINDING_SEVERITY_OPTIONS: { value: FindingSeverity; label: string }[] = [
  { value: "critical", label: "Critical" },
  { value: "high", label: "High" },
  { value: "medium", label: "Medium" },
  { value: "low", label: "Low" },
  { value: "info", label: "Informational" },
];

export const FINDING_CONFIDENCE_OPTIONS: { value: FindingConfidence; label: string }[] = [
  { value: "confirmed", label: "Confirmed" },
  { value: "high", label: "High" },
  { value: "medium", label: "Medium" },
  { value: "low", label: "Low" },
];

export const FINDING_STATUS_OPTIONS: { value: FindingStatus; label: string }[] = [
  { value: "open", label: "Open" },
  { value: "confirmed", label: "Confirmed" },
  { value: "false_positive", label: "False Positive" },
  { value: "accepted_risk", label: "Accepted Risk" },
  { value: "remediated", label: "Remediated" },
  { value: "duplicate", label: "Duplicate" },
];

export const RESOLVED_FINDING_STATUSES: FindingStatus[] = ["remediated", "false_positive", "accepted_risk", "duplicate"];

export interface FindingReference {
  id: string;
  reference_type: FindingReferenceType;
  reference_value: string;
}

export interface FindingEvidence {
  id: string;
  source_tool: string;
  title: string | null;
  content: string | null;
  file_path: string | null;
  created_at: string;
}

export interface FindingSummary {
  id: string;
  assessment_id: string;
  host_id: string | null;
  rule_id: string;
  plugin: string | null;
  title: string;
  severity: FindingSeverity;
  confidence: FindingConfidence;
  category: string | null;
  status: FindingStatus;
  first_seen: string;
  last_seen: string;
  host_label: string | null;
  evidence_count: number;
  observation_count: number;
}

export interface FindingDetail {
  id: string;
  assessment_id: string;
  host_id: string | null;
  source_execution_id: string | null;
  rule_id: string;
  plugin: string | null;
  title: string;
  description: string | null;
  impact: string | null;
  severity: FindingSeverity;
  confidence: FindingConfidence;
  category: string | null;
  cvss_score: number | null;
  cwe: string | null;
  owasp: string | null;
  remediation: string | null;
  status: FindingStatus;
  first_seen: string;
  last_seen: string;
  created_at: string;
  updated_at: string;
  host: HostSummary | null;
  affected_services: Service[];
  supporting_observations: Observation[];
  evidence: FindingEvidence[];
  references: FindingReference[];
  execution_history: ExecutionHistoryEntry[];
}

export interface FindingListParams {
  assessment_id?: string;
  host_id?: string;
  severity?: FindingSeverity;
  confidence?: FindingConfidence;
  status?: FindingStatus;
  category?: string;
  plugin?: string;
  rule_id?: string;
  search?: string;
  sort_by?: string;
  sort_dir?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

export interface CorrelationRun {
  id: string;
  assessment_id: string | null;
  status: "running" | "completed" | "failed";
  started_at: string;
  completed_at: string | null;
  hosts_evaluated: number;
  rules_evaluated: number;
  findings_created: number;
  findings_updated: number;
  error_message: string | null;
}

export interface CorrelationRunResult {
  assessment_id: string | null;
  hosts_evaluated: number;
  rules_evaluated: number;
  rule_count: number;
  findings_created: number;
  findings_updated: number;
}

export interface CorrelationStatus {
  registered_rule_count: number;
  last_run: CorrelationRun | null;
  recent_runs: CorrelationRun[];
}
